### ISSUE TITLE
'ip_forward' is disabled in the load balancer network namespace

### ISSUE THREAD

**Issue by aston-r** [moby/moby#38723]
Hi,

I faced the issues that docker container can not reach another docker container via the vip address. I investigated the issue and the issue is caused by disabled ip_forward in load balancer network namespace:
```
nsenter --net=/var/run/docker/netns/lb_nlw0vbakt cat /proc/sys/net/ipv4/ip_forward
0
```

Here is more information about my environment:
```
Bare Metal Server->LXC Container->Docker Swarm services with vip endpoint mode and overlay network

- ip_forward is disabled on Bare Metal Server;
- ip_forward is disabled in LXC container;
```

Docker Swarm uses ipvs module in masquarading mode and ip_forward should be enabled in LB namespace. 

Here is simple STR that illustrates the problem:
- disable ip_forward on Bare Metal Server;
- run lxc container with enabled ip_forward;
- create network namespace in lxc container;
- check the ip_forward value in the network namespace;
```
> cat /proc/sys/net/ipv4/ip_forward
1
> ip netns add test
> ip netns exec test cat /proc/sys/net/ipv4/ip_forward
0
```

In other words the ip_forward value is inherited from Bare Metal Server, so it order to workaround the issue I had to enable ip_forward on the Bare Metal Server too.

I tested this on debian with the latest avalaible kernel/lxc/docker-ce. 

Could you please enable ip_forward explicitly in the network namespaces that are created by docker?

**Comment by pm-melo** [comment]
Are there alternatives for this behaviour? Is this still relevant? If not, why?

**Comment by aleck011mi** [comment]
# Docker Swarm VIP Routing Broken in LXC: `ip_forward=0` in lb Namespaces

## Environment

| Component | Version |
|-----------|---------|
| Proxmox | 8.x (3-node cluster) |
| Docker | 28.5.2 (also reproduced on 29.3.0) |
| LXC | Privileged containers, nesting enabled |
| Kernel | 6.14.11-6-pve |
| Swarm | Overlay networks, VIP endpoint mode |

---

## Problem

Docker Swarm VIP routing is completely broken in privileged LXC containers. Containers can:

- ✅ Resolve service names correctly via DNS
- ✅ Ping container IPs directly
- ✅ Connect to container IPs via TCP
- ❌ Connect to VIPs — TCP connections hang indefinitely

---

## Root Cause Confirmed

`ip_forward` is `0` in all lb namespaces after Docker creates them:

```bash
$ for ns in /var/run/docker/netns/lb_*; do
    echo -n "$ns: "
    nsenter --net=$ns cat /proc/sys/net/ipv4/ip_forward
done
/var/run/docker/netns/lb_xun00u51y: 0
/var/run/docker/netns/lb_nfjjlsckm: 0
/var/run/docker/netns/lb_oibfdg17k: 0
```

Diagnosis confirmed by checking mangle table counters — IPVS **is** marking packets correctly (75+ hits on the VIP), but cannot forward them because `ip_forward=0`:

```bash
$ nsenter --net=/var/run/docker/netns/lb_xun00u51y iptables -t mangle -L -n -v

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
   75  4500 MARK       0    --  *      *       0.0.0.0/0            10.0.4.2   MARK set 0x157
```

The vxlan tunnel is healthy, FDB entries are correct, and the IPVS rules point to the right container IP — the only broken piece is `ip_forward=0` preventing IPVS from forwarding the marked packets.

---

## Confirmed NOT Fixed By

All of the following were tested and confirmed to have **no effect** on the lb namespace `ip_forward` value:

- Setting `net.ipv4.ip_forward=1` on the Proxmox host
- Setting `net.ipv4.ip_forward=1` in `/etc/sysctl.d/99-ip-forward.conf` on the Docker LXC node
- Adding `lxc.sysctl.net.ipv4.ip_forward = 1` to the LXC container config on the Proxmox host
- `net.core.devconf_inherit_init_net` — not available in kernel 6.14
- Docker daemon restart
- Full stack removal and redeployment
- `docker service update --force`

This is **not** an LXC permissions issue. The container has full capabilities:

```
CapEff: 000001ffffffffff
CapBnd: 000001ffffffffff
```

Nesting is enabled, and manual `nsenter` commands succeed. **Docker is explicitly creating lb namespaces with `ip_forward=0`** and nothing in the host or container configuration can override this.

---

## Confirmed Fix

Manually setting `ip_forward=1` in the lb namespace resolves VIP routing **immediately**:

```bash
$ nsenter --net=/var/run/docker/netns/lb_xun00u51y sysctl -w net.ipv4.ip_forward=1
net.ipv4.ip_forward = 1

$ docker run --rm --network my_overlay alpine nc -zv my_service 5432
my_service (10.0.4.2:5432) open
```

---

## Production Workaround

Since Docker recreates lb namespaces with `ip_forward=0` every time a new overlay network is deployed, a polling systemd service is required as a workaround:

**`/usr/local/bin/docker-swarm-fix.sh`**

```bash
#!/bin/bash
while true; do
    for ns in /var/run/docker/netns/lb_* /var/run/docker/netns/ingress_sbox; do
        [ -f "$ns" ] || continue
        val=$(nsenter --net=$ns cat /proc/sys/net/ipv4/ip_forward 2>/dev/null)
        if [ "$val" = "0" ]; then
            nsenter --net=$ns sysctl -w net.ipv4.ip_forward=1 2>/dev/null
        fi
    done
    sleep 10
done
```

**`/etc/systemd/system/docker-swarm-fix.service`**

```ini
[Unit]
Description=Fix ip_forward in Docker Swarm namespaces
After=docker.service
Wants=docker.service

[Service]
Type=simple
Restart=always
RestartSec=10
ExecStart=/usr/local/bin/docker-swarm-fix.sh

[Install]
WantedBy=multi-user.target
```

---

## Expected Behavior

Docker should set `net.ipv4.ip_forward=1` when creating lb namespaces. IPVS masquerading mode requires IP forwarding to be enabled — creating lb namespaces with it disabled makes VIP routing non-functional by design.

The fix belongs in the lb namespace initialization code in `libnetwork`.

---

## Impact

This issue makes Docker Swarm VIP mode completely non-functional in LXC environments, which are widely used for running Docker on Proxmox. The only workaround requires a privileged polling service running on every Swarm node, correcting a value that Docker should be setting correctly in the first place.
